File: app/javascript/controllers/pdf_preview_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["container", "loading", "image", "fallback"]
  static values = { documentId: Number }

  connect() {

    // Si pas de loading target, l'aperçu est déjà en cache - ne rien faire
    if (!this.hasLoadingTarget) {
      return
    }

    // Générer l'aperçu de manière asynchrone seulement si nécessaire
    this.generatePdfPreview()
  }

  async generatePdfPreview() {
    try {

      const response = await fetch(`/api/pdf_preview/${this.documentIdValue}/generate`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').getAttribute('content')
        }
      })

      const result = await response.json()

      if (result.success && result.preview_url) {
        // Recharger la page pour afficher l'aperçu en cache
        window.location.reload()
      } else {
        this.showFallback('Impossible de générer l\'aperçu')
      }
    } catch (error) {
      this.showFallback('Erreur de chargement')
    }
  }

  showFallback(message) {
    if (this.hasLoadingTarget) {
      this.loadingTarget.innerHTML = `
        <div class="text-center">
          <i class="bi bi-file-earmark-pdf text-danger" style="font-size: 2.5rem;"></i>
          <div class="small text-muted mt-2">${message}</div>
        </div>
      `
    }
  }
}

File: app/controllers/api/pdf_preview_controller.rb
class Api::PdfPreviewController < ApplicationController
  before_action :authenticate_user!

  def generate
    document = current_user.documents.find(params[:id])

    if document.is_pdf?
      # Génération asynchrone de l'aperçu
      preview_url = PdfPreviewService.generate_preview_for_document(document)

      if preview_url
        render json: {
          success: true,
          preview_url: preview_url,
          document_id: document.id
        }
      else
        render json: {
          success: false,
          error: "Impossible de générer l'aperçu PDF",
          document_id: document.id
        }, status: 422
      end
    else
      render json: {
        success: false,
        error: "Le document n'est pas un PDF",
        document_id: document.id
      }, status: 400
    end
  rescue ActiveRecord::RecordNotFound
    render json: {
      success: false,
      error: "Document non trouvé"
    }, status: 404
  rescue => e
    Rails.logger.error "Erreur génération aperçu PDF: #{e.message}"
    render json: {
      success: false,
      error: "Erreur serveur",
      document_id: params[:id]
    }, status: 500
  end
end
